x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/allocation/AllocationState.java

```diff
@@ -7,16 +7,23 @@
 
 package org.elasticsearch.xpack.core.ml.inference.allocation;
 
+import java.util.Arrays;
 import java.util.Locale;
 
 public enum AllocationState {
+    STARTING,
+    PARTIALLY_STARTED,
```

Comment:
I am worried about the semantics of `partially_started`. I think it can lead to confusion as to whether the user should proceed running inference or not against the model. I wonder if we should distinguish the state here, which is useful to tell us whether the allocation has started or whether it is stopping, from the allocation health, which is whether the model has been allocated to zero nodes, some nodes, all eligible nodes. We could even follow the index health paradigm and have that take values of `red, yellow, green`.

Reply:
I think another benefit of this is that we don't have a new state to update. I would hope we can calculate health on the fly when the stats API is called.

Reply:
I have received push-back from folks for `red, yellow, green` (I originally thought this best...and still do). If you have a better name, please let me know.

@joshdevins @davidkyle ^^

> I think another benefit of this is that we don't have a new state to update. I would hope we can calculate health on the fly when the stats API is called.

I was initially doing this, but it didn't really reduce complexity all that much, I can revisit.

Reply:
I like `red, yellow, green` because it will be intuitive to people who are used to looking at Elasticsearch index status.  What were the arguments against copying that approach?  Maybe they were good enough arguments to justify not using `red, yellow, green` here, but we should make sure they're written down as somebody will ask why we didn't use `red, yellow, green` in the future.

Reply:
`red, yellow, green` is only meaningful if you are familiar with the cluster health API. There is some general intuition from traffic lights that green is go and red stop but we should strive for something more informative so users don't have to look up exactly what `yellow` means.

`partially_deployed` is clearer to me.

Reply:
> `partially_deployed` is clearer to me.

It should be `partially_allocated` then, because we're doing model allocations, with deployment being a higher level thing.

And for consistency we should change `starting` to `not_allocated` and `started` to `fully_allocated`, so we have `not_allocated -> partially_allocated -> fully_allocated`.  `stopping` is more tricky but maybe could be `deallocating`.

Reply:
+1

`not_allocated ` -> `unallocated`

or 

`starting` -> `allocating` 

as technically a stopped deployment isn't allocated

Reply:
@davidkyle @droberts195 

It is definitely tricky to have a "started" and a "partially_started" type of thing. 

What if we added a separate "health" status that is NOT persisted in cluster state?

The allocation state will stay as "stopped, started, starting..." with no indication of health. Started just means there is at least one node that has the model loaded. 

And a new generated value field of `allocation_health` is added? As for the enumerations in `allocation_health`, I am open for suggestions.

Reply:
I'm thinking about where I would use a health status. Usually I want to start a deployment and throw data at it either through the _infer API or an ingest processor. When something goes wrong and I start seeing errors the first place I will look is the stats API where I will be checking for failed nodes, not nodes the nodes that never hosted the model. `partially_allocated` does not tell me if a deployment contains nodes that once worked and have now failed so there is an argument for a health status - "warning: deployment X has nodes [Y, Z] in a failed state with reason [blah]".

However, I'm not sure we can reliably detect a node which is part of a deployment has failed as opposed to a node that has left the cluster or a node that never hosted the model. There is engineering work to do to get there and that is beyond the scope of this PR but I agree some kind of health reporting would be very nice.

Reply:
> The allocation state will stay as "stopped, started, starting..." with no indication of health. Started just means there is at least one node that has the model loaded.

I think this could be a good idea.

As the PR currently stands it's reimplementing the logic of `red, yellow, green` and `wait for yellow status` as they apply to index health, but is _not_ reusing the exact same terms, so people won't immediately assume we've copied the same logic.  If we are not copying the exact terminology then we have the chance to do something different.

The original comment on this thread got lost in the subsequent discussion.  The word "partial" is confusing to users, because they'll have to look in the docs to find out if "partial" means "usable" or "you need to wait a bit longer".  We didn't want to use `red, yellow, green` on the grounds that would force people to read the docs, so we shouldn't use the word "partial" for the same reason.  So if we want to stick with the semantics of `red, yellow, green`, but change the names then maybe `starting, started, fully_allocated` would be better names.  I also think the default for the `waitFor` argument on start should be the thing that corresponds to yellow if we stick with those semantics, as that's when the model is usable.

But even though it's relatively easy to rename the enum values before release we should definitely think more about whether we want the enum to contain separate values for partial and complete, because that's more of a far-reaching change.  In the future will anyone want to say something like, "I want this model on 6 nodes and wait for it to be on 2 nodes before I use it"?  That would imply `waitFor` should be a number rather than a state, perhaps with a magic `_all` value also allowed.  And the enum just needs the `starting`, `started`, `stopping` values it has today.  It's clear that `started` means "possible to use".  And then some other field(s) indicate how many nodes the model is supposed to be on and how many it's actually on.

Reply:
@droberts195 My latest commit adds a new field to the stats called `health` and its `starting` `started` `fully_allocated`. 

The `waitFor` parameter now references that enumeration as well. 

NOTE: Health is only returned in the allocation stats object if the state is `starting` or `started`.